# Exploring Hacker News Posts

Hacker News is a site started by the startup incubator Y Combinator, where user-submitted stories (known as "posts") receive votes and comments, similar to reddit. Hacker News is extremely popular in technology and startup circles, and posts that make it to the top of the Hacker News listings can get hundreds of thousands of visitors as a result.

[You can find the data set here](https://www.kaggle.com/datasets/hacker-news/hacker-news-posts), but note that we have reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that didn't receive any comments and then randomly sampling from the remaining submissions. 

**Below are descriptions of the columns:**
- id: the unique identifier from Hacker News for the post
- title: the title of the post
- url: the URL that the posts links to, if the post has a URL
- num_points: the number of points the post acquired, calculated as the total number of upvotes minus the total number of downvotes
- num_comments: the number of comments on the post
- author: the username of the person who submitted the post
- created_at: the date and time of the post's submission

_Read the hacker_news.csv file in as a list of lists, assign the result to the variable hn._

In [140]:
opened_file = open('hacker_news.csv')
from csv import reader
read_file = reader(opened_file)
hn = list(read_file)
header = hn[0:1]
hn = hn[1:]

_Display the first five rows of hn_

In [141]:
print(header,'\n')

for x in hn[:5]:
    print(x)
    print('\n')

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']] 

['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52']


['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30']


['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20']


['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']


['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']




## Removing Headers from a List of Lists

_In order to analyze our data, we need to first remove the row containing the column headers. Let's extract the first row of data, and assign it to the variable headers._

In [142]:
headers = hn[0]

_Remove the first row from hn_

In [143]:
hn = hn[1:]

In [144]:
print(headers)

['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52']


In [145]:
print(hn[0:5])

[['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12'], ['10482257', 'Title II kills investment? Comcast and other ISPs are now spending more', 'http://arstechnica.com/business/2015/10/comcast-and-other-isps-boost-network-investment-despite-net-neutrality/', '53', '22', 'Deinos

## Extracting "Ask HN" and "Show HN" Posts

_Now that we've removed the headers from hn, we're ready to filter our data. Since we're only concerned with post titles beginning with Ask HN or Show HN, we'll create new lists of lists containing just the data for those titles._

_Create three empty lists called ask_posts, show_posts, and other_posts._

In [146]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    title = title.lower()
    if title.startswith('ask hn'):
        ask_posts.append(row)        
    elif title.startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)

#Lets check the number of posts in ask_posts, show_posts, and other_posts:
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
17193


## Calculating the Average Number of Comments for Ask HN and Show HN Posts

_Next we'll determine if Ask Posts or Show Posts receive more comments on average._

In [147]:
total_ask_comment = 0

for row in ask_posts:
    num_comments = int(row[4])
    total_ask_comment += num_comments

avg_ask_comments = total_ask_comment / len(ask_posts)

print(avg_ask_comments)

total_show_comment = 0

for row in show_posts:
    num_comments = int(row[4])
    total_show_comment += num_comments

avg_show_comments = total_show_comment / len(show_posts)

print(avg_show_comments)

14.038417431192661
10.31669535283993


##  Finding the Number of Ask Posts and Comments by Hour Created

_On average, ask posts receive more comments than show posts. Since ask posts are more likely to receive comments, we'll focus our remaining analysis just on these posts._

_Next we'll calculate the number of ask posts created in each hour of the day, along with the number of comments received._

In [148]:
import datetime as dt

result_list = []

for row in ask_posts:
    result_list.append(
        [row[6], int(row[4])]
    )

counts_by_hour = {}
comments_by_hour = {}
date_format = "%m/%d/%Y %H:%M"

for row in result_list:
    date = row[0]
    comments = row[1]
    time = dt.datetime.strptime(date, date_format).strftime("%H")
    if time not in counts_by_hour:
        counts_by_hour[time] = 1
        comments_by_hour[time] = comments
    else:
        counts_by_hour[time] += 1
        comments_by_hour[time] += comments

comments_by_hour

{'09': 251,
 '13': 1253,
 '10': 793,
 '14': 1416,
 '16': 1814,
 '23': 543,
 '12': 687,
 '17': 1146,
 '15': 4477,
 '21': 1745,
 '20': 1722,
 '02': 1381,
 '18': 1439,
 '03': 421,
 '05': 464,
 '19': 1188,
 '01': 683,
 '22': 479,
 '08': 492,
 '04': 337,
 '00': 447,
 '06': 397,
 '07': 267,
 '11': 641}

## Calculating the Average Number of Comments for Ask HN Posts by Hour


_On the previous screen, we created two dictionaries:_

- **counts_by_hour:** _contains the number of ask posts created during each hour of the day._
- **comments_by_hour:** _contains the corresponding number of comments ask posts created at each hour received._

_Below we calculate the average number of comments for posts created during each hour of the day_

In [149]:
avg_by_hour = []

for hour in comments_by_hour:
    avg_by_hour.append(
        [hour, comments_by_hour[hour]/counts_by_hour[hour]]
    )
    

avg_by_hour

[['09', 5.5777777777777775],
 ['13', 14.741176470588234],
 ['10', 13.440677966101696],
 ['14', 13.233644859813085],
 ['16', 16.796296296296298],
 ['23', 7.985294117647059],
 ['12', 9.41095890410959],
 ['17', 11.46],
 ['15', 38.5948275862069],
 ['21', 16.009174311926607],
 ['20', 21.525],
 ['02', 23.810344827586206],
 ['18', 13.20183486238532],
 ['03', 7.796296296296297],
 ['05', 10.08695652173913],
 ['19', 10.8],
 ['01', 11.383333333333333],
 ['22', 6.746478873239437],
 ['08', 10.25],
 ['04', 7.170212765957447],
 ['00', 8.127272727272727],
 ['06', 9.022727272727273],
 ['07', 7.852941176470588],
 ['11', 11.051724137931034]]

In [150]:
swap_avg_by_hour = []

for row in avg_by_hour:
    swap_avg_by_hour.append(
        [row[1], row[0]]
    )

swap_avg_by_hour

[[5.5777777777777775, '09'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [16.796296296296298, '16'],
 [7.985294117647059, '23'],
 [9.41095890410959, '12'],
 [11.46, '17'],
 [38.5948275862069, '15'],
 [16.009174311926607, '21'],
 [21.525, '20'],
 [23.810344827586206, '02'],
 [13.20183486238532, '18'],
 [7.796296296296297, '03'],
 [10.08695652173913, '05'],
 [10.8, '19'],
 [11.383333333333333, '01'],
 [6.746478873239437, '22'],
 [10.25, '08'],
 [7.170212765957447, '04'],
 [8.127272727272727, '00'],
 [9.022727272727273, '06'],
 [7.852941176470588, '07'],
 [11.051724137931034, '11']]

_Although we now have the results we need, this format makes it difficult to identify the hours with the highest values. Let's finish by sorting the list of lists and printing the five highest values in a format that's easier to read._

In [151]:
sorted_swap = sorted(swap_avg_by_hour, reverse=True)
print("Top 5 Hours for Ask Posts Comments")
print("\n")
import datetime as dt
time_format = "%H" #15

for row in sorted_swap[:5]:
    average = "{:.2f}".format(row[0])
    hour = row[1]
    new_hour = dt.datetime.strptime(hour, time_format)
    print("{hr}:00 {avg} average comments per post".format(hr=hour, avg=average))

Top 5 Hours for Ask Posts Comments


15:00 38.59 average comments per post
02:00 23.81 average comments per post
20:00 21.52 average comments per post
16:00 16.80 average comments per post
21:00 16.01 average comments per post


Based on my findings posts submitted 15:00 has by far a higher chance of receiving comments, with an average of 38.59 comments per post it outwins the second runner up (02:00) by over 60%.

According to [the data set documentation](https://www.kaggle.com/datasets/hacker-news/hacker-news-posts), the timezone used is Eastern Time in the US. Since I am located in UTC that would mean that the peak hour for posting is 20:00 local time.